File: R/cmb_table.R
#' @name CmbTable-class
#'
#' @aliases
#' Rcpp_CmbTable Rcpp_CmbTable-class CmbTable
#'
#' @title Class for counting unique combinations of integers
#'
#' @description
#' `CmbTable` implements a hash table having a vector of integers as the key,
#' and the count of occurrences of each unique integer combination as the
#' value. A unique ID is assigned to each unique combination of input values.
#'
#' `CmbTable` is a C++ class exposed directly to R (via `RCPP_EXPOSED_CLASS`).
#' Methods of the class are accessed using the `$` operator. **Note that all
#' arguments to class methods are required and must be given in the
#' order documented.** Naming the arguments is optional but may be preferred
#' for readability.
#'
#' @param keyLen The number of integer values comprising each combination.
#' @param varNames Optional character vector of names for the variables in the
#' combination.
#' @returns An object of class `CmbTable`. Contains a hash table having a
#' vector of `keyLen` integers as the key, and the count of occurrences of each
#' unique integer combination as the value. Class methods that operate on the
#' hash table are described in Details.
#'
#' @section Usage (see Details):
#' \preformatted{
#' ## Constructors
#' cmb <- new(CmbTable, keyLen)
#' # or, giving the variable names:
#' cmb <- new(CmbTable, keyLen, varNames)
#'
#' ## Methods
#' cmb$update(int_cmb, incr)
#' cmb$updateFromMatrix(int_cmbs, incr)
#' cmb$updateFromMatrixByRow(int_cmbs, incr)
#' cmb$asDataFrame()
#' cmb$asMatrix()
#' }
#'
#' @section Details:
#' ## Constructors
#'
#' \code{new(CmbTable, keyLen)}\cr
#' Default variable names will be assigned as `V1`, `V2`, ....
#' Returns an object of class `CmbTable`.
#'
#' \code{new(CmbTable, keyLen, varNames)}\cr
#' Alternate constructor to specify variable names.
#' Returns an object of class `CmbTable`.
#'
#' ## Methods
#'
#' \code{$update(int_cmb, incr)}\cr
#' Updates the hash table for the integer combination in the numeric vector
#' `int_cmb` (coerced to integer by truncation).
#' If this combination exists in the table, its count will be
#' incremented by `incr`. If the combination is not found in the table,
#' it will be inserted with count set to `incr`.
#' Returns the unique ID assigned to this combination.
#' Combination IDs are sequential integers starting at `1`.
#'
#' \code{$updateFromMatrix(int_cmbs, incr)}\cr
#' This method is the same as `$update()` but for a numeric matrix of
#' integer combinations `int_cmbs` (coerced to integer by truncation).
#' The matrix is arranged with each column vector forming an integer
#' combination. For example, the rows of the matrix could be
#' one row each from a set of `keyLen` rasters all read at the
#' same extent and pixel resolution (i.e., row-by-row raster overlay).
#' The method calls `$update()` on each combination (each column of
#' `int_cmbs`), incrementing count by `incr` for existing
#' combinations, or inserting new combinations with count set to `incr`.
#' Returns a numeric vector of length `ncol(int_cmbs)` containing the
#' IDs assigned to the combinations.
#'
#' \code{$updateFromMatrixByRow(int_cmbs, incr)}\cr
#' This method is the same as `$updateFromMatrix()` above except the
#' integer combinations are in rows of the matrix `int_cmbs` (columns
#' are the variables).
#' The method calls `$update()` on each combination (each row of
#' `int_cmbs`), incrementing count by `incr` for existing
#' combinations, or inserting new combinations with count set to `incr`.
#' Returns a numeric vector of length `nrow(int_cmbs)` containing the
#' IDs assigned to the combinations.
#'
#' \code{$asDataFrame()}\cr
#' Returns the `CmbTable` as a data frame with column `cmbid` containing
#' the unique combination IDs, column `count` containing the counts of
#' occurrences, and `keyLen` columns (with names from `varNames`) containing
#' the integer values comprising each unique combination.
#'
#' \code{$asMatrix()}\cr
#' Returns the `CmbTable` as a matrix with column `1` (`cmbid`)
#' containing the unique combination IDs, column `2` (`count`)
#' containing the counts of occurrences, and columns `3:keyLen+2`
#' (with names from `varNames`) containing the integer values comprising each
#' unique combination.
#'
#' @examples
#' m <- matrix(c(1,2,3,1,2,3,4,5,6,1,3,2,4,5,6,1,1,1), 3, 6, byrow = FALSE)
#' rownames(m) <- c("layer1", "layer2", "layer3")
#' print(m)
#'
#' cmb <- new(CmbTable, 3, rownames(m))
#' cmb
#'
#' cmb$updateFromMatrix(m, 1)
#' cmb$asDataFrame()
#'
#' cmb$update(c(4,5,6), 1)
#' cmb$update(c(1,3,5), 1)
#' cmb$asDataFrame()
#'
#' # same as above but matrix arranged with integer combinations in the rows
#' m <- matrix(c(1,2,3,1,2,3,4,5,6,1,3,2,4,5,6,1,1,1), 6, 3, byrow = TRUE)
#' colnames(m) <- c("V1", "V2", "V3")
#' print(m)
#'
#' cmb <- new(CmbTable, 3)
#' cmb$updateFromMatrixByRow(m, 1)
#' cmb$asDataFrame()
#'
#' cmb$update(c(4,5,6), 1)
#' cmb$update(c(1,3,5), 1)
#' cmb$asDataFrame()
NULL

Rcpp::loadModule("mod_cmb_table", TRUE)

File: src/cmb_table.cpp
/* Implementation of class CmbTable
   Chris Toney <chris.toney at usda.gov>
*/

#include "cmb_table.h"

#include <Rcpp.h>

#include <string>
#include <vector>

#include "rcpp_util.h"

CmbTable::CmbTable()
        : CmbTable(1, Rcpp::CharacterVector::create()) {}

CmbTable::CmbTable(int keyLen)
        : CmbTable(keyLen, Rcpp::CharacterVector::create()) {}

CmbTable::CmbTable(int keyLen, const Rcpp::CharacterVector &varNames)
        : m_key_len(keyLen),
          m_var_names(Rcpp::as<std::vector<std::string>>(varNames))  {

    if (keyLen <= 0)
        Rcpp::stop("'keyLen' must be a positive integer");

    m_key_len = static_cast<R_xlen_t>(keyLen);

    if (varNames.size() == 0) {
        for (R_xlen_t i = 0; i < m_key_len; ++i) {
            Rcpp::String s = "V" + std::to_string(i + 1);
            m_var_names.push_back(s);
        }
    }

    if (m_key_len != static_cast<R_xlen_t>(m_var_names.size()))
        Rcpp::stop("'keyLen' must equal 'length(varNames)'");
}

double CmbTable::update(const Rcpp::IntegerVector &int_cmb, double incr) {
    // Increment count for existing int_cmb
    // or insert new int_cmb with count = incr.
    cmbKey key(int_cmb);
    cmbData &cmbdat = m_cmb_map[key];
    cmbdat.count += incr;
    if (cmbdat.ID == 0.0) {
        m_last_ID += 1.0;
        cmbdat.ID = m_last_ID;
    }
    return cmbdat.ID;
}

Rcpp::NumericVector CmbTable::updateFromMatrix(
    const Rcpp::IntegerMatrix &int_cmbs, double incr) {

    // int combinations (int_cmbs) are columns of a matrix (nrow = m_key_len).
    // Increment count for existing int_cmb,
    // else insert new int_cmb with count = incr.
    // Return a vector of cmb IDs for the columns of the input matrix.
    if (int_cmbs.nrow() != m_key_len) {
        Rcpp::stop("number of matrix rows must equal the key length: " +
                   std::to_string(m_key_len));
    }

    const R_xlen_t ncol = int_cmbs.ncol();
    Rcpp::NumericVector out(ncol);

    for (R_xlen_t k = 0; k < ncol; ++k) {
        out[k] = update(int_cmbs.column(k), incr);
    }
    return out;
}

Rcpp::NumericVector CmbTable::updateFromMatrixByRow(
    const Rcpp::IntegerMatrix &int_cmbs, double incr) {

    // int combinations (int_cmbs) are rows of a matrix (ncol = m_key_len).
    // Same as updateFromMatrix() except by row instead of by column
    // (i.e., variables here are in the columns).
    // Return a vector of cmb IDs for the rows of the input matrix.
    if (int_cmbs.ncol() != m_key_len) {
        Rcpp::stop("number of matrix columns must equal the key length: " +
                   std::to_string(m_key_len));
    }

    const R_xlen_t nrow = int_cmbs.nrow();
    Rcpp::NumericVector out(nrow);

    for (R_xlen_t k = 0; k < nrow; ++k) {
        out[k] = update(int_cmbs.row(k), incr);
    }
    return out;
}

Rcpp::DataFrame CmbTable::asDataFrame() const {
    Rcpp::NumericVector dvCmbID(m_cmb_map.size());
    Rcpp::NumericVector dvCmbCount(m_cmb_map.size());
    std::vector<Rcpp::IntegerVector> aVec(m_key_len);
    cmbKey key;
    cmbData cmbdat;

    for (R_xlen_t i = 0; i < m_key_len; ++i) {
        aVec[i] = Rcpp::IntegerVector(m_cmb_map.size());
    }
    std::size_t this_idx = 0;
    for (auto iter = m_cmb_map.begin(); iter != m_cmb_map.end(); ++iter) {
        key = iter->first;
        cmbdat = iter->second;
        dvCmbID[this_idx] = cmbdat.ID;
        dvCmbCount[this_idx] = cmbdat.count;
        for (R_xlen_t var = 0; var < m_key_len; ++var) {
            aVec[var][this_idx] = key.cmb[var];
        }
        ++this_idx;
    }

    Rcpp::DataFrame dfOut = Rcpp::DataFrame::create();
    dfOut.push_back(dvCmbID, "cmbid");
    dfOut.push_back(dvCmbCount, "count");
    for (R_xlen_t i = 0; i < m_key_len; ++i) {
        dfOut.push_back(aVec[i], m_var_names[i]);
    }
    return dfOut;
}

Rcpp::NumericMatrix CmbTable::asMatrix() const {
    Rcpp::NumericMatrix m_out = df_to_matrix_(asDataFrame());

    Rcpp::CharacterVector cvColNames = Rcpp::clone(Rcpp::wrap(m_var_names));

    cvColNames.push_front("count");
    cvColNames.push_front("cmbid");
    Rcpp::colnames(m_out) = cvColNames;

    return m_out;
}

void CmbTable::show() const {
    std::string out = "cmbid count";
    for (const auto& s : m_var_names) {
        out += (" " + s);
    }
    Rcpp::Rcout << "C++ object of class CmbTable\n";
    Rcpp::Rcout << " Columns: " << out.c_str() << "\n";
}

RCPP_MODULE(mod_cmb_table) {
    Rcpp::class_<CmbTable>("CmbTable")

    .constructor
        ("Default constructor (combination vector of length 1)")
    .constructor<int>
        ("Length of the combination vector")
    .constructor<int, Rcpp::CharacterVector>
        ("Length of the combination vector, vector of variable names")

    .method("update", &CmbTable::update,
        "Increment by incr if int_cmb exists, else insert with count = incr")
    .method("updateFromMatrix", &CmbTable::updateFromMatrix,
        "update() on integer combinations contained in columns of a matrix")
    .method("updateFromMatrixByRow", &CmbTable::updateFromMatrixByRow,
        "update() on integer combinations contained in rows of a matrix")
    .const_method("asDataFrame", &CmbTable::asDataFrame,
        "Returns a dataframe containing the combinations table")
    .const_method("asMatrix", &CmbTable::asMatrix,
        "Returns a matrix containing the combinations table")
    .const_method("show", &CmbTable::show,
        "S4 show()")
    ;
}

